### ISSUE TITLE
'Protobuf compiler version doesn't match library version' only when using manifest mode

### ISSUE THREAD

**Issue by Fenrirthviti** [microsoft/vcpkg#14902]
Environment

Windows 10
CMake 3.16.2
Visual Studio 2019
vcpkg on commit (62fe6ff)

This might be related to #7157, but as that issue has long since been closed I am creating a new issue.

I am getting a similar issue on my end with protobuf, and none of the fixes in #7157 seem to correct the issue. Frustratingly, the issue only seems to crop up when using a vcpkg.json in manifest mode. Even more frustrating, is that simply running the cmake command again causes everything to complete successfully. I've been banging my head against this for almost a week now, any assistance or direction is welcome.

```
$ cmake -B build64 -S .
-- Building for: Visual Studio 16 2019
-- vcpkg install may take a few minutes, please be patient
-- Running vcpkg install
-- Running vcpkg install - done
-- The C compiler identification is MSVC 19.28.29334.0
-- The CXX compiler identification is MSVC 19.28.29334.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.28.29333/bin/Hostx64/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- CPACK_PACKAGE_VERSION:
-- COCKATRICE_VERSION: 2020-11-23-Development-2.7.6-beta.2-10-ga107c7c4-modified: Blessed Sanctuary
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
-- UPDATE TRANSLATIONS: OFF
-- Found Qt 5.15.2
CMake Warning at C:/Program Files/CMake/share/cmake-3.16/Modules/FindProtobuf.cmake:499 (message):
  Protobuf compiler version doesn't match library version 3.13.0
Call Stack (most recent call first):
  build64/vcpkg_installed/x64-windows/share/protobuf/vcpkg-cmake-wrapper.cmake:13 (_find_package)
  vcpkg/scripts/buildsystems/vcpkg.cmake:573 (include)
  CMakeLists.txt:213 (FIND_PACKAGE)


-- Found Protobuf: optimized;F:/Source/Repos/Cockatrice/build64/vcpkg_installed/x64-windows/lib/libprotobuf.lib;debug;F:/Source/Repos/Cockatrice/build64/vcpkg_installed/x64-windows/debug/lib/libprotobufd.lib (found version "3.13.0")
CMake Error at CMakeLists.txt:215 (MESSAGE):
  No protoc command found!


-- Configuring incomplete, errors occurred!
See also "F:/Source/Repos/Cockatrice/build64/CMakeFiles/CMakeOutput.log".
See also "F:/Source/Repos/Cockatrice/build64/CMakeFiles/CMakeError.log".
```

To clarify, when running vcpkg manually using a response file (or just specifying the package), and then running cmake after, everything works properly. This is only an issue on the *first* run with vcpkg in manifest mode.

Edit: As it might be relevant, here is the failing check in the project's top-level [CMakeLists.txt](https://github.com/Fenrirthviti/Cockatrice/blob/cmake-updates/CMakeLists.txt):

```cmake
# Find other needed libraries
FIND_PACKAGE(Protobuf REQUIRED)
IF(NOT EXISTS "${Protobuf_PROTOC_EXECUTABLE}")
  MESSAGE(FATAL_ERROR "No protoc command found!")
ENDIF()
```

**Comment by JackBoosY** [comment]
Should related to the warning message:
```
CMake Warning at C:/Program Files/CMake/share/cmake-3.16/Modules/FindProtobuf.cmake:499 (message):
  Protobuf compiler version doesn't match library version 3.13.0
```

**Comment by Fenrirthviti** [comment]
I've checked all the installed files, and they are all 3.13.0, so this error doesn't make any sense in context. Any insight on why this is only happening with manifest mode run as part of cmake, and doesn't happen when running vcpkg separately?

Additionally, this happens in a vacuum in a clean VM with nothing pre-installed (such as a CI instance).

**Comment by JackBoosY** [comment]
Nope. will try to repro this issue tomorrow.

**Comment by JackBoosY** [comment]
I can't repro this issue:
```
1> -- Found Protobuf: optimized;F:/vcpkg/installed/x86-windows/lib/libprotobuf.lib;debug;F:/vcpkg/installed/x86-windows/debug/lib/libprotobufd.lib (found version "3.14.0") 
1> Protobuf_PROTOC_EXECUTABLE: F:/vcpkg/installed/x86-windows/tools/protobuf/protoc.exe
```
Did you installed cmake manually and its version beyond 3.18?

**Comment by Fenrirthviti** [comment]
I am using cmake 3.16 locally, and 3.18 is being used in CI. The issue exists in both places. I am fully willing to admit this might be an issue with the project, but this is repeatable 100% of the time from a completely clean setup using the Cockatrice project. Are you using the same commit that I am for vcpkg? I see that your version is finding 3.14 where mine is finding 3.13. I will also try updating vcpkg locally and see if this corrects the issue.

**Comment by Fenrirthviti** [comment]
I can confirm updating vcpkg to latest master solves this issue, but it is unclear exactly what caused it in the first place. Feel free to close, but if anyone else runs in to this issue, an update to vcpkg (as a submodule in my case) seems to have resolved it.

**Comment by Fenrirthviti** [comment]
Actually, that worked locally, but not in CI, it's still failing here:

https://github.com/Fenrirthviti/Cockatrice/runs/1498534574?check_suite_focus=true

**Comment by Fenrirthviti** [comment]
I've also replicated locally again as well:

```
CMake Warning at C:/Program Files/CMake/share/cmake-3.16/Modules/FindProtobuf.cmake:499 (message):
  Protobuf compiler version doesn't match library version 3.14.0
Call Stack (most recent call first):
  build64/vcpkg_installed/x64-windows/share/protobuf/vcpkg-cmake-wrapper.cmake:13 (_find_package)
  vcpkg/scripts/buildsystems/vcpkg.cmake:573 (include)
  CMakeLists.txt:190 (FIND_PACKAGE)


Found Protobuf: optimized;F:/Source/Repos/Cockatrice/build64/vcpkg_installed/x64-windows/lib/libprotobuf.lib;debug;F:/Source/Repos/Cockatrice/build64/vcpkg_installed/x64-windows/debug/lib/libprotobufd.lib (found version "3.14.0") 
CMake Error at CMakeLists.txt:192 (MESSAGE):
  No protoc command found!


Configuring incomplete, errors occurred!
```

I tested again in a clean VM with no prior installations. This issue is extremely frustrating because on any subsequent runs past the first, it runs without issue. Something is either not being properly set until much later in the cmake workflow. I am really not sure, which makes troubleshooting this very difficult to tell if it's actually fixed or not.

**Comment by Fenrirthviti** [comment]
Please let me know if there is any additional information that I can provide to help isolate what the issue is here. If you could share what you did to test locally or if you could share that project so I can test and compare with the Cockatrice project, that would be helpful as well.

Thank you in advance.

**Comment by JackBoosY** [comment]
@Fenrirthviti I don't know... it may be caused by different versions of cmake.

**Comment by Fenrirthviti** [comment]
Thanks for the reply, I have tested on 3.16, 3.18, and 3.19 which all exhibit the same issue. To confirm, you were able to successfully configure the Cockatrice project with no errors? Or what exactly was your test? There may be answers in how you set up your test that I can use to solve this issue in my project.

**Comment by JackBoosY** [comment]
Even cannot repro on Visual Studio 2019:
```
1> CMake generation started for configuration: 'x86-Release'.
1> The toolchain file has changed (CMAKE_TOOLCHAIN_FILE).
1> Command line: "cmd.exe" /c ""C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Visual Studio 16 2019" -A Win32  -DCMAKE_CONFIGURATION_TYPES:STRING="Release" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\usr\source\repos\CMakeProject2\out\install\x86-Release" -DCMAKE_TOOLCHAIN_FILE:FILEPATH="F:/vcpkg/scripts/buildsystems/vcpkg.cmake"  "C:\Users\usr\source\repos\CMakeProject2" 2>&1"
1> Working directory: C:\Users\usr\source\repos\CMakeProject2\out\build\x86-Release
1> [CMake] -- Selecting Windows SDK version 10.0.18362.0 to target Windows 10.0.19041.
1> [CMake] -- The C compiler identification is MSVC 19.27.29111.0
1> [CMake] -- The CXX compiler identification is MSVC 19.27.29111.0
1> [CMake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x86/cl.exe
1> [CMake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x86/cl.exe - works
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - done
1> [CMake] -- Detecting C compile features
1> [CMake] -- Detecting C compile features - done
1> [CMake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x86/cl.exe
1> [CMake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx64/x86/cl.exe - works
1> [CMake] -- Detecting CXX compiler ABI info
1> [CMake] -- Detecting CXX compiler ABI info - done
1> [CMake] -- Detecting CXX compile features
1> [CMake] -- Detecting CXX compile features - done
1> [CMake] -- Found Protobuf: optimized;F:/vcpkg/installed/x86-windows/lib/libprotobuf.lib;debug;F:/vcpkg/installed/x86-windows/debug/lib/libprotobufd.lib (found version "3.14.0") 
1> [CMake] Protobuf_PROTOC_EXECUTABLE: F:/vcpkg/installed/x86-windows/tools/protobuf/protoc.exe
1> [CMake] -- Configuring done
1> [CMake] -- Generating done
1> [CMake] -- Build files have been written to: C:/Users/usr/source/repos/CMakeProject2/out/build/x86-Release
1> Extracted CMake variables.
1> Extracted source files and headers.
1> Extracted code model.
1> Extracted includes paths.
1> CMake generation finished.
```

**Comment by Fenrirthviti** [comment]
Grab my branch here: <https://github.com/Fenrirthviti/Cockatrice/tree/vcpkg-updates>

Qt is required, any 5.x version (currently 5.12.9 is preferred), then just run:

```
mkdir build
cd build
cmake .. -DQTDIR="<path to qt msvc folder>"
```

**Comment by JackBoosY** [comment]
Currently blocked on the qt5 build:
```
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_IDXGIFactory2
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3D11DeviceContext1
qrhid3d11.obj : error LNK2001: unresolved external symbol IID_ID3DUserDefinedAnnotation
..\..\lib\Qt5Guid.dll : fatal error LNK1120: 3 unresolved externals
```

Needs Neumann-A's PR merge first.

**Comment by JackBoosY** [comment]
I noticed that you may use the binary cache, could you remove the file which printed on the install command?
Like `Stored binary cache: C:\Users\usr\AppData\Local\vcpkg\archives\47\476d5dec5e744c15484259ffba1481106d2dfb86.zip`.

**Comment by Fenrirthviti** [comment]
I have been clearing those caches from appdata with every run.

I can also replicate this in a Windows Sandbox environment from scratch on first run. I am happy to record a video showing my exact steps that I am following if this might be useful. They are currently:

- Install VS2019 Community
- Install Git
- Install Qt from qi.io (not using vcpkg version of Qt as it generally takes 2-3 times longer to install via vcpkg)
- Run cmake steps from my previous comment
- Observe the configure fails

I have also had several other friends/coworkers test this process on their systems and they are all able to replicate. I do not understand what you are doing differently, or have configured differently on your system, that is causing you to not run in to this issue. Can you please provide the entire test cmake project you used and the exact complete full steps you are following, including every single exact command so I can compare on my side?

It is extremely frustrating that I have replicated this on no less than 4 different machines on my end, and no less than 6 other people have replicated it on their end. This is also replicated in CI, per the link I provided earlier in the conversation.

Again, I am absolutely willing to entertain that I am just doing something fundamentally wrong when it comes to vcpkg, but I have been trying to solve this for almost 2 weeks now and have made no progress. I'll ask again, PLEASE provide your test project so I can run it and see if the same issue occurs and compare what might be happening on my end.

**Comment by JackBoosY** [comment]
cc @BillyONeal can you please take a look?

I'm using the Visual Studio 2019 cmake project. Here is the sample:
CMakeLists.txt:
```cmake
# CMakeList.txt : CMake project for CMakeProject2, include source and define
# project specific logic here.
#
cmake_minimum_required (VERSION 3.8)

# Add source to this project's executable.
add_executable (CMakeProject2 "CMakeProject2.cpp" "CMakeProject2.h")

FIND_PACKAGE(Protobuf REQUIRED)
IF(NOT EXISTS "${Protobuf_PROTOC_EXECUTABLE}")
  MESSAGE(FATAL_ERROR "No protoc command found!")
ENDIF()
message("Protobuf_PROTOC_EXECUTABLE: ${Protobuf_PROTOC_EXECUTABLE}")
```

CMakeSettings.json:
```json
{
  "configurations": [
    {
      "name": "x86-Release",
      "generator": "Visual Studio 16 2019",
      "configurationType": "Release",
      "inheritEnvironments": [ "msvc_x86" ],
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "",
      "ctestCommandArgs": "",
      "variables": [
        {
          "name": "CMAKE_TOOLCHAIN_FILE",
          "value": "F:/vcpkg/scripts/buildsystems/vcpkg.cmake",
          "type": "FILEPATH"
        }
      ]
    }
  ]
}
```
CMakeProject2.cpp:
```cpp
// CMakeProject2.cpp : Defines the entry point for the application.
//

#include "CMakeProject2.h"

int main(int argc, char **argv)
{
	return 0;
}
```
CMakeProject2.h:
```cpp
// CMakeProject2.h : Include file for standard system include files,
// or project specific include files.

#pragma once

#include <iostream>
```

**Comment by Fenrirthviti** [comment]
Thanks, that definitely raises a few questions on my side that might explain why I am running in to issues. My vcpkg.json is following the example from the docs, and only contains:

```json
{
  "name": "cockatrice",
  "version-string": "0.0.0-dev",
  "dependencies": [
    "openssl",
    "protobuf",
    "liblzma",
    "zlib",
    "gtest"
  ]
}
```

But I see that your CMakeSettings.json is far more detailed, but also doesn't specify the dependencies. Are all of those options required to use vcpkg in manifest mode? Or are some missing from mine which might be contributing to this issue? Is it automatically detecting what dependencies it needs to download based on `find_package()` statements in the cmakelists? I can't seem to find clear answers in the documentation for this offhand, so any direction here is appreciated as well.

And thank you for the time in helping me solve this.

**Comment by JackBoosY** [comment]
@Fenrirthviti  No, `CMakeSettings.json` is only used to set cmake options, not for manifest.

**Comment by BillyONeal** [comment]
> cc @BillyONeal can you please take a look?
> 

What's your `vcpkg.json` look like?

**Comment by JackBoosY** [comment]
@BillyONeal https://github.com/microsoft/vcpkg/issues/14902#issuecomment-741961222 here.

**Comment by Fenrirthviti** [comment]
Correct, the whole branch is also located here if you would like to check anything else: https://github.com/Fenrirthviti/Cockatrice/tree/vcpkg-updates

**Comment by strega-nil** [comment]
hmm, this is really odd. My only thought is... no, I really have no idea...

**Comment by strega-nil** [comment]
I will look into this Soon (TM)

**Comment by Fenrirthviti** [comment]
@strega-nil Thank you! I am available pretty much any time if you need any further information or logs from my end.

**Comment by strega-nil** [comment]
@Fenrirthviti are you on discord? it might be easier to figure this out with a more synchronous discussion system.

**Comment by Fenrirthviti** [comment]
I am, yes, `Len#9999`.

**Comment by Fenrirthviti** [comment]
@strega-nil Just wanted to check in, I should be available pretty much any time at the above alias. Thanks!

**Comment by strega-nil** [comment]
@Fenrirthviti sorry, it's been christmas in the US, so I've been out of the office with my family 😄 

I'll ping you in the new year if that's okay?

**Comment by Fenrirthviti** [comment]
No worries at all, speak then!

**Comment by Fenrirthviti** [comment]
Just a poke at this. It looks like the vcpkg.txt methods (which the Cockatrice project is using as a workaround) have been removed in later versions, so the manifest mode being broken is bit of a challenge.

**Comment by strega-nil** [comment]
Ah, thanks @Fenrirthviti; I'll take a look.

**Comment by strega-nil** [comment]
@strega-nil could you ping me on discord?

**Comment by JackBoosY** [comment]
Does this issue resolved?

**Comment by Fenrirthviti** [comment]
Nobody ever reached out to me regarding this issue, I have still been waiting. I have not tested in some time as I have not had time, but as far as I know this is still an issue.

**Comment by dg0yt** [comment]
@JackBoosY  Did you actually use manifest mode when trying to reproduce the issue? 
 `Protobuf_PROTOC_EXECUTABLE: F:/vcpkg/installed/x86-windows/tools/protobuf/protoc.exe` indicates a classic mode installation. Manifest mode should use a path which contains `vcpkg_installed` (as found in the comments by @Fenrirthviti).

**Comment by JackBoosY** [comment]
Ping @strega-nil-ms

**Comment by strega-nil-ms** [comment]
I'm looking at this rn

**Comment by strega-nil-ms** [comment]
Hmm, it seems that I can't repro it; `git rev-parse HEAD`: `c79e8b4c624104fefa3dd81ec7cf6499cc9d030e` of the cockatrice repo works for me. Is there any more information?

(the vcpkg repo is at `dc9e8e1497ee60f65c21928eca5eb5fd5cce69f3`)

**Comment by Fenrirthviti** [comment]
I will revisit this when time allows this week and update as soon as I can.

**Comment by Fenrirthviti** [comment]
I've validate that this error does not seem to occur locally with latest master of the vcpkg repo, but I will need additional time to test in CI to validate it's properly fixed there as well.

**Comment by JackBoosY** [comment]
We haven't been able to repro this; if more information comes up, or this issue appears again, please reopen.

**Comment by Fenrirthviti** [comment]
No worries, it was replicable 100% on my system, so if it's still an issue and I end up having time to go back to that project I will follow up, I have just not had the time to devote to setting this build environment back up. Thanks everyone for their time and efforts.
